File: Sources/AgentCoding/Mitm/SecretsVault.swift
import Foundation
import CryptoKit
import Security

/// Per-install master key + symmetric encryption helpers for
/// at-rest secret storage. The master key is a 32-byte random AES-256
/// stored in the macOS **Data Protection Keychain** — the iOS-style
/// keychain that's app-scoped by code-signing identity and (critically)
/// never prompts the user for access. The legacy "file-based" macOS
/// keychain is what would pop the "Always Allow / Allow / Deny" dialog
/// every time the app rebuilt; DPKC sidesteps that entirely.
///
/// Wiping the keychain entry rotates the key — every existing secret
/// blob becomes unreadable, but the non-sensitive metadata in
/// profile.json is unaffected, so the user can re-enter their API keys.
public enum SecretsVault {
    /// Service tag for the master-key keychain item. Versioned so we
    /// can rotate the storage scheme without overwriting user data.
    private static let service = "io.bromure.agentic-coding.master-key"
    private static let account = "v1"

    /// In-memory cache so we hit the keychain at most once per process.
    /// Reset on app quit. SymmetricKey holds bytes in a buffer that
    /// CryptoKit zeroes on dealloc.
    private static var cachedKey: SymmetricKey?
    private static let cacheLock = NSLock()

    /// Fetch (or lazily create) the per-install master key.
    public static func masterKey() throws -> SymmetricKey {
        cacheLock.lock(); defer { cacheLock.unlock() }
        if let k = cachedKey { return k }

        // Preferred: the Data Protection Keychain (at-rest protected, no prompt).
        if let fetched = (try? fetchFromKeychain()) ?? nil {
            cachedKey = fetched
            return fetched
        }
        let fresh = SymmetricKey(size: .bits256)
        if (try? storeInKeychain(fresh)) != nil {
            cachedKey = fresh
            return fresh
        }

        // Fallback: a 0600 key file under Application Support. The Data
        // Protection Keychain is unreachable to a signed-but-not-notarized /
        // no-provisioning-profile build, which *silently* broke every at-rest
        // feature — notably trace-body capture (encrypt threw, writeBody skipped
        // the write, and the inspector showed "not captured"). The file key is
        // weaker (it sits beside the ciphertext), but it keeps the feature
        // working; a fully-provisioned release still uses the keychain.
        let key = try fileFallbackKey()
        FileHandle.standardError.write(Data(
            "[secrets] keychain unavailable — using file-backed master key (trace/secret encryption is local-key only)\n".utf8))
        cachedKey = key
        return key
    }

    /// Master key in a 0600 file under Application Support — the fallback when
    /// the keychain is unavailable.
    private static func fileFallbackKey() throws -> SymmetricKey {
        let url = FileManager.default
            .urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("BromureAC", isDirectory: true)
            .appendingPathComponent("secrets-master.key")
        if let data = try? Data(contentsOf: url), data.count == 32 {
            return SymmetricKey(data: data)
        }
        let fresh = SymmetricKey(size: .bits256)
        let bytes: Data = fresh.withUnsafeBytes { Data($0) }
        try FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        try bytes.write(to: url, options: .atomic)
        try? FileManager.default.setAttributes(
            [.posixPermissions: NSNumber(value: 0o600)], ofItemAtPath: url.path)
        return fresh
    }

    /// Encrypt arbitrary bytes with the master key. Output is the
    /// AES-GCM "combined" form: 12-byte nonce || ciphertext || 16-byte tag.
    public static func encrypt(_ plain: Data) throws -> Data {
        let key = try masterKey()
        let sealed = try AES.GCM.seal(plain, using: key)
        guard let combined = sealed.combined else {
            throw VaultError.sealFailed
        }
        return combined
    }

    public static func decrypt(_ blob: Data) throws -> Data {
        let key = try masterKey()
        let box = try AES.GCM.SealedBox(combined: blob)
        return try AES.GCM.open(box, using: key)
    }

    // MARK: - Keychain plumbing

    private static func fetchFromKeychain() throws -> SymmetricKey? {
        let query: [CFString: Any] = [
            kSecClass:                       kSecClassGenericPassword,
            kSecAttrService:                 service,
            kSecAttrAccount:                 account,
            kSecReturnData:                  true,
            kSecMatchLimit:                  kSecMatchLimitOne,
            // CRITICAL: opt into the data-protection (iOS-style)
            // keychain on macOS so we never get the legacy "Always
            // Allow" prompt. App-scoped by signing identity; no UI.
            kSecUseDataProtectionKeychain:   true,
        ]
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        if status == errSecItemNotFound { return nil }
        guard status == errSecSuccess,
              let data = result as? Data,
              data.count == 32 else {
            throw VaultError.keychain("fetch", status)
        }
        return SymmetricKey(data: data)
    }

    private static func storeInKeychain(_ key: SymmetricKey) throws {
        let data: Data = key.withUnsafeBytes { Data($0) }
        let attrs: [CFString: Any] = [
            kSecClass:                     kSecClassGenericPassword,
            kSecAttrService:               service,
            kSecAttrAccount:               account,
            kSecValueData:                 data,
            // Available after the user has unlocked the device once
            // since boot — same window the user uses bromure-ac in.
            // ThisDeviceOnly prevents iCloud sync so secrets never
            // leave this Mac.
            kSecAttrAccessible:            kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
            kSecUseDataProtectionKeychain: true,
        ]
        let status = SecItemAdd(attrs as CFDictionary, nil)
        if status == errSecSuccess { return }
        if status == errSecDuplicateItem {
            // Race: another concurrent call won. Re-fetch and use that.
            return
        }
        throw VaultError.keychain("add", status)
    }
}

public enum VaultError: Error, CustomStringConvertible {
    case keychain(String, OSStatus)
    case sealFailed

    public var description: String {
        switch self {
        case .keychain(let op, let s): return "Keychain \(op) failed (\(s))"
        case .sealFailed:              return "AES-GCM seal returned no combined output"
        }
    }
}

File: windows/Bromure.AC.Mitm/Vault/SecretsVault.cs
using System.Security.Cryptography;
using Bromure.AC.Mitm.Trace;
using Bromure.Platform;

namespace Bromure.AC.Mitm.Vault;

/// <summary>
/// Direct port of <c>Sources/AgentCoding/Mitm/SecretsVault.swift</c>.
/// Per-install AES-256 master key + AES-GCM encrypt/decrypt helpers for
/// at-rest secret storage.
///
/// <para>On macOS the master key lives in the Data Protection Keychain
/// (app-scoped, no prompts). On Windows it's a 32-byte random key
/// stored as a DPAPI-LocalMachine blob via <see cref="ISecretStore"/>,
/// which mirrors the same threat model: bound to this device, app-
/// scoped via the DPAPI entropy, no user interaction.</para>
///
/// <para>Wiping the blob (or the install dir + uninstall) rotates the
/// key. Every existing encrypted blob becomes unreadable, but the
/// non-sensitive metadata in profile.json is unaffected — the user
/// can re-enter their API keys.</para>
/// </summary>
public sealed class SecretsVault : IBodyEncryptor
{
    private const string KeyBlobName = "vault-master-key-v1";
    private readonly ISecretStore _store;
    private readonly object _gate = new();
    private byte[]? _cachedKey;

    public SecretsVault(ISecretStore store) => _store = store;

    /// <summary>Fetch (or lazily create) the per-install master key.</summary>
    public byte[] MasterKey()
    {
        lock (_gate)
        {
            if (_cachedKey is not null) return _cachedKey;

            var existing = _store.ReadBlob(KeyBlobName, BlobScope.LocalMachine);
            if (existing is { Length: 32 })
            {
                _cachedKey = existing;
                return existing;
            }
            var fresh = RandomNumberGenerator.GetBytes(32);
            _store.StoreBlob(KeyBlobName, fresh, BlobScope.LocalMachine);
            _cachedKey = fresh;
            return fresh;
        }
    }

    /// <summary>
    /// Encrypt with the master key. Output layout matches macOS:
    /// 12-byte nonce || ciphertext || 16-byte tag (the AES-GCM "combined"
    /// form). This makes encrypted blobs interchangeable with macOS
    /// builds when both unwrap the same master key — which they don't,
    /// but the layout is the right one to keep so a future cross-host
    /// sync doesn't have to re-encode.
    /// </summary>
    public byte[] Encrypt(ReadOnlySpan<byte> plaintext)
    {
        var key = MasterKey();
        var nonce = RandomNumberGenerator.GetBytes(12);
        var ciphertext = new byte[plaintext.Length];
        var tag = new byte[16];
        using var aes = new AesGcm(key, tag.Length);
        aes.Encrypt(nonce, plaintext, ciphertext, tag);

        var output = new byte[nonce.Length + ciphertext.Length + tag.Length];
        Buffer.BlockCopy(nonce, 0, output, 0, nonce.Length);
        Buffer.BlockCopy(ciphertext, 0, output, nonce.Length, ciphertext.Length);
        Buffer.BlockCopy(tag, 0, output, nonce.Length + ciphertext.Length, tag.Length);
        return output;
    }

    public byte[] Decrypt(ReadOnlySpan<byte> blob)
    {
        if (blob.Length < 12 + 16) throw new VaultException("blob too short for nonce + tag");
        var key = MasterKey();
        var nonce = blob[..12].ToArray();
        var tag = blob[(blob.Length - 16)..].ToArray();
        var ciphertext = blob[12..(blob.Length - 16)].ToArray();
        var plaintext = new byte[ciphertext.Length];
        using var aes = new AesGcm(key, tag.Length);
        aes.Decrypt(nonce, ciphertext, tag, plaintext);
        return plaintext;
    }

    /// <summary>Forget the cached key (next call re-reads from <see cref="ISecretStore"/>).</summary>
    public void ClearCache()
    {
        lock (_gate) _cachedKey = null;
    }
}

public sealed class VaultException : Exception
{
    public VaultException(string message) : base(message) { }
}
